File: Python3/leet1178/Solution.py
from typing import List

class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        first_cond = []
        for word in words:
            first_cond.append([])
            first_cond[-1] = [False] * 26
            for w in word:
                first_cond[-1][ord(w) - ord('a')] = True

        res = [0] * len(puzzles)

        for i, p in enumerate(puzzles):
            table = [False] * 26
            for v in p:
                table[ord(v) - ord('a')] = True

            for j, w in enumerate(words):
                if first_cond[j][ord(p[0]) - ord('a')] == False:
                    continue

                key = True
                for k in range(26):
                    if first_cond[j][k] and (not table[k]):
                        key = False
                        break

                if key:
                    res[i] += 1

        return res

File: C++/leet1178/Solution.cpp
//
// Created by YUANYIN on 2021/02/26.
//

#include "../leetcode.h"

class Solution {
public:
    vector<int> findNumOfValidWords(vector<string>& words, vector<string>& puzzles) {
        unordered_map<int, int> freq;
        for (const string& word : words) {
            int mask = 0;
            for (char ch : word) {
                mask |= (1 << (ch - 'a'));
            }
            if (__builtin_popcount(mask) <= 7) {
                ++ freq[mask];
            }
        }

        vector<int> ans;
        for (const string& puzzle : puzzles) {
            int total = 0;
            int mask = 0;
            for (int i = 1; i < 7; i++) {
                mask |= (1 << (puzzle[i] - 'a'));
            }
            int subset = mask;
            do {
                int s = subset | (1 << (puzzle[0] - 'a'));
                if (freq.find(s) != freq.end()) {
                    total += freq[s];
                }
                subset = (subset - 1) & mask;
            } while (subset != mask);

            ans.push_back(total);
        }
        return ans;
    }
};